### PR TITLE
Introduce "High Quality" Systemd unit files for `rosenpass` and `rp`

### PR DESCRIPTION
**Pull request by tfc**
This PR introduces systemd unit files to simplify running `rosenpass` and`rp` with minimal setup. The workflow after these changes is like this:

1. Create configuration files, private/public keys on both server and client, and distribute them accordingly
2. Start the service using `systemctl start rp@<config-name>.service` (or respectively `rosenpass@<config-name>.service`)
3. Enjoy a post-quantum-secure encrypted connection

I tested this manually on an Ubuntu machine. The automated tests use NixOS, see next section.

## Changes Introduced

To achieve the above, the following changes have been made:

- Rosenpass Unit File: 
  - Added a systemd unit file for `rosenpass`
  - Implemented a NixOS integration test for verification

- `rp` App: Patched the `rp` application to support configuration files

- `rp` Unit File:
  - Added a systemd unit file for `rp`
  - Added a NixOS integration test for verification

## Considerations and Notes

While working on this, a few challenges and limitations were encountered:

- Rust Code Quality: The Rust code may not be best practice and could be improved. Feedback on how to refine it is appreciated.
- `rosenpass` Client Config Parsing: The `listen` field is currently required in the client configuration, which feels wrong. However, i feel addressing this would be feature-creep in this PR.
- `rp` IP Parameter Limitation: The `ip` parameter for `rp exchange` could theoretically support multiple IPs, but this functionality was not implemented in this iteration, as it already supports the standard scenarios.

The need to extend the `rp` code was not anticipated during the initial planning of this work package. However, the current implementation is functional for simple use cases and should provide a solid foundation for further enhancements if needed.

I hope this implementation aligns with your goals and expectations. Feedback and suggestions for improvement are welcome!

### CODE REVIEW

**Review by koraa — COMMENTED** [review]
Looks pretty good!

**Comment by koraa** [comment]
Darwin (the system underlying mac) is failing; something about the futures crate.

We don't really support darwin/mac right now so I do not believe making sure everything works perfectly in darwin is worth the effort here. However, we do need to make the CI pass…

File: systemd/rp@.service
Comment (tfc) [reply systemd/rp@.service]: I screwed up my rebase. My fault. It is all in order now.

File: tests/systemd/rp.nix
Comment (tfc) [reply tests/systemd/rp.nix:1]: Yes. The tests are both installed in the flake `checks` which means they are at least run by `nix flake check`

**Comment by tfc** [comment]
> Darwin (the system underlying mac) is failing; something about the futures crate.
> 
> [..] However, we do need to make the CI pass…

Isn't that then a problem in all the other PRs, too?

File: tests/systemd/rp.nix
Comment (koraa) [reply tests/systemd/rp.nix:1]: Thanks!

**Comment by koraa** [comment]
> > Darwin (the system underlying mac) is failing; something about the futures crate.
> > [..] However, we do need to make the CI pass…
> 
> Isn't that then a problem in all the other PRs, too?

Well, the other PRs haven't triggered the problem recently.

**Comment by koraa** [comment]
> > Well, the other PRs haven't triggered the problem recently.
> 
> Looking at the error messages, it seems like both rosenpass and rp trigger this. I wonder why only in this PR, as it only patches rp and not rosenpass.
> 
> I also looked at how to fix it and make it build on my macbook, which was successful, however looking at policy decisions like [this one](https://github.com/rosenpass/rosenpass/blob/main/rp/Cargo.toml#L27) without the knowledge which platforms _are_ supported, i am not sure if the `futures = ...` line should simply be pulled out of the conditional block or if there is a better solution.

Honestly, I have no idea. The `futures` crate is not used outside of the `rp` subcrate.

I do not see a compelling reason to not pull futures out of the block; the rp subcrate was not written to the same quality standard as the rest of rosenpass, so I am not too surprised about some noise.

> > We don't really support darwin/mac right now so I do not believe making sure everything works perfectly in darwin is worth the effort here. However, we do need to make the CI pass…
> 
> I wonder why is darwin then even activated in the CI? Is that by accident?

Someone contributed a feature and then abandoned it. We haven't had enough compelling reasons to throw it out again.

**Comment by tfc** [comment]
@koraa Thanks for looking into the PR!

I’ve pushed a fix that makes it build on my M3 MacBook (aarch64). However, the following CI jobs are still failing:

- Nix / Build x86_64-darwin.rosenpass
- Nix / Build x86_64-darwin.rp
- QC / cargo-test (macos-13)

The first two are clearly for x86_64. For the third one, I went through the logs but couldn’t figure out for sure if it’s also x86_64. It seems likely, though.

Unfortunately, I don’t have an x86_64 Mac to debug this. From what I can see in the logs, it looks like the `futures` dependency isn’t being used on darwin-x86_64, even though I changed it to apply to all platforms.

Since macOS isn’t a main target for the project, and fixing this would take extra time and tools I don’t have right now, I feel like this is out of scope for my work here. I’m usually happy to help beyond what I’m directly responsible for, but this issue is starting to block the tasks I was actually hired to do.

If someone with the right setup can take a closer look, I’m happy to help however I can to move this forward. Let me know how we can proceed!

**Comment by koraa** [comment]
@tfc Thank you for your work.

I took a look at the issue and the problem was luckily easy to fix by adding the `futures` crate to the list of dependencies unconditionally.

https://github.com/rosenpass/rosenpass/pull/521/commits/6b3bebb9b9b2326f8ce13a4a1b9e4be26b2576c2

Whereas previously, the futures crate was used in a block enabled on freebsd and linux exclusively

https://github.com/rosenpass/rosenpass/blob/c9da9b85913b8747e9965eaae30390f089cbec47/rp/src/exchange.rs#L35-L42

this PR introduced an unconditional import of the futures crate. The issues could be resolved by adding the futures crate to the list of dependencies, but since the futures crate was used only in the platform specific code block, I was able to resolve the issue without resorting to a non-platform specific code by adding appropriate platform feature flags to the pieces of code using the imports.

In other words, this was not a platform specific issue and merely a case of incorrectly configured platform specific code.

Can you please cherry-pick my fix into your branch (I can not push to yours) and rebase the branch?

@aparcar wanted to manually give this a whirl before merge, but other than that I think this PR is ready for merge.

Thank you for the contribution, this has become an excellent update to make `rp` more useful!

**Comment by tfc** [comment]
@koraa i cherrypicked the commit. Please approve the PR workflows so that we can see how it works.

File: rp/src/cli.rs
Comment (aparcar) [inline rp/src/cli.rs:156]: This should be "ip option [...]" I reckon

File: tests/systemd/rp.nix
Comment (aparcar) [inline tests/systemd/rp.nix:64]: Could you please explain these lines? Isn't it just a "fancy" `cp` and `rm`? If so I'd prefer to be more *verbose* and use those commands. Maybe a matter of taste. (same for `rosenpass.nix`)

Comment (tfc) [reply tests/systemd/rp.nix:64]: Do you see the `sed` commands inbetween, that substitute all kinds of substrings? Can you please elaborate how you would you do that with `cp` and `rm`?

File: rp/src/cli.rs
Comment (tfc) [reply rp/src/cli.rs]: fixed

File: tests/systemd/rp.nix
Comment (aparcar) [reply tests/systemd/rp.nix:64]: Thanks, I misread that (still catching up on nix overall). Just for clarification, the first line only copies the contents right? 
```            
< ${pkgs.rosenpass}/lib/systemd/system/rosenpass.target > $out/lib/systemd/system/rosenpass.target
```

Comment (tfc) [reply tests/systemd/rp.nix:64]: The first line just performs a plain copy, because in this file, no substitution happens.

The second file's content needs a lot of substitution.

Both lines simply happen in the "same style", which is of course a matter of taste.

However, this shell script snippet has nothing to do with nix inherently (apart from the ${...} string interpolation), it's pure bashism.

Comment (aparcar) [reply tests/systemd/rp.nix:64]: Thanks!

File: tests/systemd/rosenpass.nix
Comment (aparcar) [inline tests/systemd/rosenpass.nix:42]:
```suggestion
    listen = [ ];
```

I just tested this and it works, too. Right now, `rosenpass` excepts it to be there, but if it's empty, [it performs some magic on it's own](https://github.com/rosenpass/rosenpass/issues/487).

**Comment by aparcar** [comment]
I think this is really nice and especially the NixOS based configuration of two VMs running Rosenpass server/client quite impressive!

Since I barely touched NixOS prior to this, could you please give a minimal example on how this would look like in real life? Could you kindly give me some pointers? I got a VM running NixOS, how would I install Rosenpass with your changes (i.e. something in the lines of `nixos-install --flake https://github.com/rosenpass/rosenpass`?) and then configure it to connect to my hosts Rosenpass instance? I'd like to document this somehow since this seems super helpful for reproducible setups, however each NixOS guide I find suggests somewhat different steps (a fast moving target, I assume).

File: tests/systemd/rosenpass.nix
Comment (tfc) [reply tests/systemd/rosenpass.nix]: Thank you! Fixed.

**Comment by tfc** [comment]
> I think this is really nice and especially the NixOS based configuration of two VMs running Rosenpass server/client quite impressive!
> 
> Since I barely touched NixOS prior to this, could you please give a minimal example on how this would look like in real life? Could you kindly give me some pointers? I got a VM running NixOS, how would I install Rosenpass with your changes (i.e. something in the lines of `nixos-install --flake https://github.com/rosenpass/rosenpass`?) and then configure it to connect to my hosts Rosenpass instance? I'd like to document this somehow since this seems super helpful for reproducible setups, however each NixOS guide I find suggests somewhat different steps (a fast moving target, I assume).

This workpackage has *nothing* to do with NixOS. The work order was to create systemd unit files that work on Ubuntu, Debian, Fedora, et. al. when they come packaged with the binary.

This is *not* the prefered way to set rosenpass or rp up on NixOS, as these already exist upstream in the nixpkgs repo.
The NixOS test solely proves that the new systemd unit files work well.
Tests that automate tests in legacy distros are planned but out of scope of this PR.

**Comment by aparcar** [comment]
Thanks for this clarification, I'll merge this once all tests passed.

**Review by aparcar — APPROVED** [review]

**Comment by koraa** [comment]
@aparcar Thank you! Good work!
@tfc You too of course.

@aparcar I am trying to start being consistent about how I merge PRs…I am now rebasing first and then I produce a merge commit that could be fast forward…just to document which commits belonged to which PR

**Pull request merged**
